File: 383. Ransom Note/main.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        count = [0] * 26

        for c in magazine:
            count[ord(c) - ord("a")] += 1
        for c in ransomNote:
            count[ord(c) - ord("a")] -= 1
            if count[ord(c) - ord("a")] < 0:
                return False

        return True


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        dict = {}

        for c in magazine:
            if c not in dict:
                dict[c] = 1
            else:
                dict[c] += 1
        for c in ransomNote:
            if c in dict and dict[c] > 0:
                dict[c] -= 1
            else:
                return False

        return True


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        for c in ransomNote:
            if c in magazine:
                magazine = magazine.replace(c, "", 1)
            else:
                return False

        return True

File: 383. Ransom Note/main.rs
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut count = [0; 26];
        magazine
            .chars()
            .for_each(|c| count[c as usize - 'a' as usize] += 1);
        !ransom_note.chars().any(|c| {
            count[c as usize - 'a' as usize] -= 1;
            count[c as usize - 'a' as usize] < 0
        })
    }
}

use std::collections::HashMap;

impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut hashmap = HashMap::new();
        magazine.chars().for_each(|c| {
            hashmap.entry(c).and_modify(|val| *val += 1).or_insert(1);
        });
        for c in ransom_note.chars() {
            match hashmap.get_mut(&c) {
                Some(n) if *n > 0 => *n -= 1,
                _ => return false,
            }
        }

        true
    }
}
